internal/cli/status.go

```diff
@@ -0,0 +1,850 @@
+
+	// Determine which view to show based on user input
+	var projectTarget, appTarget string
+	if len(cmdArgs) >= 1 {
+		match := reAppTarget.FindStringSubmatch(cmdArgs[0])
+
+		if match != nil {
+			projectTarget = match[1]
+			appTarget = match[2]
+		} else {
+			projectTarget = cmdArgs[0]
+		}
+	} else if len(cmdArgs) == 0 {
+		// If we're in a project dir, load the name. Otherwise we'll
+		// show a list of all projects and their status and leave projectTarget
+		// blank
+		if c.project.Ref() != nil {
+			projectTarget = c.project.Ref().Project
+		}
+	}
+
+	if appTarget == "" && c.flagApp != "" {
+		appTarget = c.flagApp
+	} else if appTarget != "" && c.flagApp != "" {
+		// setting app target and passing the flag app is a collision
+		c.ui.Output(wpAppFlagAndTargetIncludedMsg, terminal.WithWarningStyle())
+	}
+
+	// Generate a status view
+	if projectTarget == "" || c.flagAllProjects {
+		// Show high-level status of all projects
+		err = c.FormatProjectStatus()
+		if err != nil {
+			c.ui.Output("CLI failed to build project statuses: "+clierrors.Humanize(err), terminal.WithErrorStyle())
+			return 1
+		}
+	} else if projectTarget != "" && appTarget == "" {
+		// Show status of apps inside project
+		err = c.FormatProjectAppStatus(projectTarget)
+		if err != nil {
+			if status.Code(err) == codes.NotFound {
+				c.ui.Output(wpProjectNotFound, projectTarget, c.serverCtx.Server.Address, terminal.WithErrorStyle())
+			} else {
+				c.ui.Output("CLI failed to format project app statuses:"+clierrors.Humanize(err), terminal.WithErrorStyle())
+			}
+			return 1
+		}
+	} else if projectTarget != "" && appTarget != "" {
+		// Advanced view of a single app status
+		err = c.FormatAppStatus(projectTarget, appTarget)
+		if err != nil {
+			if status.Code(err) == codes.NotFound {
+				c.ui.Output(wpAppNotFound, appTarget, projectTarget, c.serverCtx.Server.Address, terminal.WithErrorStyle())
+			} else {
+				c.ui.Output("CLI failed to format app status:"+clierrors.Humanize(err), terminal.WithErrorStyle())
+			}
+			return 1
+		}
+	}
+
+	return 0
+}
+
+// FormatProjectAppStatus formats all applications inside a project
+func (c *StatusCommand) FormatProjectAppStatus(projectTarget string) error {
+	if !c.flagJson {
+		c.ui.Output(wpStatusProjectMsg, projectTarget, c.serverCtx.Server.Address)
+	}
+
+	// Get our API client
+	client := c.project.Client()
+
+	resp, err := client.GetProject(c.Ctx, &pb.GetProjectRequest{
+		Project: &pb.Ref_Project{
+			Project: projectTarget,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	project := resp.Project
+
+	workspace, err := c.getWorkspaceFromProject(resp)
+	if err != nil {
+		return err
+	}
+
+	// Summary
+	//   App list
+
+	appHeaders := []string{
+		"App", "Workspace", "Latest Status", "Last Check",
+	}
+
+	appTbl := terminal.NewTable(appHeaders...)
+
+	appFailures := false
+	for _, app := range project.Applications {
+		appStatusResp, err := client.GetLatestStatusReport(c.Ctx, &pb.GetLatestStatusReportRequest{
```

Comment:
That's expected in this case I think. The column we're displaying for this app is the _latest status_ check recorded rather than an aggregate of its status history.

Reply:
Hmm, this deviates from the logic in the RFC:

> For the top level status command, we want to let users know the health of a project given all of its applications health. When we determine the value of this column, we’ll need to look through each application's overall health status and show the total statuses for each app. We will omit any status state that isn’t covered. For example, if we have 3 apps, 2 of which are READY and 1 is error, we should show “2 READY 1 ERROR” for the overall aggregate app status.

I think I much prefer the RFC logic. As an app developer, I don't think this field is useful to me if it's flapping between new and old deployments and releases. I would expect this to be some kind of view into the health of my application.

With the bundle APIs introduced in https://github.com/hashicorp/waypoint/pull/1994, it should be pretty easy to get the latest status report for the latest deploy, the latest status report for the latest release, and then either display them both, or one combined status that can be `PARTIAL`.

Reply:
We can just display them both. I'll wait to update it once the UI bundle methods are available.